### PR TITLE
Improved shadow fade

### PR DESCRIPTION
**Pull request by lukaschod**
#### Also I decided to refactor shadow cascade GUI for HDRP and URP. For several reasons:

Partition sizes inconsistant

Last cascade had extra pixel at end inconsistantly

Any resizing was causing pixel jumping

Did not wanted to increase code duplication as extending URP cascade would resulted in very similar code of HDRP one.
I could not used HDRP solution as I had to move it to renderpipeline public core. HDRP solution was not written to be public code.

#### New shadow cascade GUI

Includes:
- Pixel perfect.
- Supports hovering.
- Supports focusing and keyboard arrow based value change.
- Support hidding border handles.
- New high quality snatch handle images.
- Gradient no longer recreates textures and uses preloaded texture.

(New urp upper, old urp bottom)

(New hdrp upper, old hdrp bottom)

## More info https://docs.google.com/document/d/1MAFk4AINTloYLgw8vBzhpah031LismehPQqVxiUhSx0.

---
### Testing status
- [x] Compare URP, HDRP and builtin shadow fade.
- [x] Test if it fixes original issues.
- [x] Test with multiple lights.
- [x] Ran gfx tests locally.
- [ ] Performance comparison. There is high chance for small regression on old mobiles as I had to add additional instruction. Sadly nothing can be done. However with baked shadow, I expect it to perform better as I got rid of one inline condition.
- [ ] Test if shadowmask correctly shadow blends to realtime shadow in all cases.
- [ ] Test if shadow fade does not start to happen in near plane, because we are no longer fading from camera, but from last cascade (I have special code path to take care of it).
- [ ] Test if new shadow cascade UI performs correctly in all cases.

---
### TODO
- [ ] Update shadow cascade images in urp docs
- [ ] Update shadow cascade images in hdrp docs
- [x] Update graphics test templates

---
### Comments to reviewers
I am mostly concerned about UI side if I did not screw up there with images, paths and so on (As rarely touch UI code).

### CODE REVIEW

File: com.unity.render-pipelines.universal/Editor/EditorUtils.cs [hunk 1]
```diff
@@ -33,207 +32,5 @@ internal class Styles
             //Measurements
             public static float defaultLineSpace = EditorGUIUtility.singleLineHeight + EditorGUIUtility.standardVerticalSpacing;
         }
-
-        static float[] splitZero = new float[0];
-        static float[] splitOne = new float[1];
-        static float[] splitTwo = new float[2];
-        static float[] splitThree = new float[3];
-        static int splitCount;
-        public static void DrawCascadeSplitGUI<T>(ref SerializedProperty shadowCascadeSplit, float distance, int cascadeCount, Unit unit)
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Editor/EditorUtils.cs:42]: As this is inside internal class, I am removing it

File: com.unity.render-pipelines.universal/Editor/Deprecated.cs [hunk 2]
```diff
@@ -15,207 +15,5 @@ namespace UnityEditor.Rendering.Universal
...
-                        shadowCascadeSplit.vector3Value = updatedValue;
-                    }
-                }
-            }
-        }
-    }
-    static partial class ShadowCascadeSplitGUI
-    {
-        [Obsolete("This is obsolete, please use HandleCascadeSliderGUI(ref float[] normalizedCascadePartitions, float distance, EditorUtils.Unit unit) instead.", false)]
-        public static void HandleCascadeSliderGUI(ref float[] normalizedCascadePartitions)
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Editor/Deprecated.cs:60]: As this is inside internal class, I am removing it

File: com.unity.render-pipelines.universal/Editor/Deprecated.cs [hunk 3]
```diff
@@ -15,207 +15,5 @@ namespace UnityEditor.Rendering.Universal
 {
     static partial class EditorUtils
     {
-        [Obsolete("This is obsolete, please use DrawCascadeSplitGUI<T>(ref SerializedProperty shadowCascadeSplit, float distance, int cascadeCount, Unit unit) instead.", false)]
-        public static void DrawCascadeSplitGUI<T>(ref SerializedProperty shadowCascadeSplit)
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Editor/Deprecated.cs:19]: As this is inside internal class, I am removing it

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 4]
```diff
@@ -214,10 +221,16 @@ void SetupMainLightShadowReceiverConstants(CommandBuffer cmd, VisibleLight shado
...
             float invHalfShadowAtlasHeight = 0.5f * invShadowAtlasHeight;
+            float softShadowsProp = softShadows ? 1.0f : 0.0f;
+
+            Vector4 lastCascade = m_CascadeSplitDistances[m_ShadowCasterCascadesCount - 1];
+            float lastCascadeRadiusSq = lastCascade.w * lastCascade.w;
+            ShadowUtils.GetScaleAndBiasForLinearDistanceFade(lastCascadeRadiusSq, m_CascadeBorder, out float shadowFadeScale, out float shadowFadeBias);
 
             cmd.SetGlobalTexture(m_MainLightShadowmap.id, m_MainLightShadowmapTexture);
             cmd.SetGlobalMatrixArray(MainLightShadowConstantBuffer._WorldToShadow, m_MainLightShadowMatrices);
-            ShadowUtils.SetupShadowReceiverConstantBuffer(cmd, m_MainLightShadowParams);
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:220]: This is no longer needed, as additional light calculation no longer uses main light

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 5]
```diff
@@ -382,18 +406,11 @@ half MainLightShadow(float4 shadowCoord, float3 positionWS, half4 shadowMask, ha
...
     half shadowFade = 1.0h;
 #endif
 
-#if defined(_MAIN_LIGHT_SHADOWS_CASCADE) && defined(CALCULATE_BAKED_SHADOWS)
-    // shadowCoord.w represents shadow cascade index
-    // in case we are out of shadow cascade we need to set shadow fade to 1.0 for correct blending
-    // it is needed when realtime shadows gets cut to early during fade and causes disconnect between baked shadow
-    shadowFade = shadowCoord.w == 4 ? 1.0h : shadowFade;
-#endif
-
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:396]: I tested, it is no longer needed as we are calculating shadow fade from cascade

This was the problem.
image

**Comment by sebastienlagarde** [comment]
Adding @RSlysz for the cascade widget part in hdrp

**Comment by Wilfrid-Unity** [comment]
Thanks for unifying the interfaces.

The refactoring for AdditionalLights look good. 
Did you check that the addition of new uniforms in Shadows.hlsl does not make shaders exceed the low GL_MAX_FRAGMENT_UNIFORM_VECTORS limit we had in some of the SHADER_API_GLES30 devices (could be as low as 224 on Adreno3xx device I think)?

I'm also adding @Verasl, who instigated previous [fixes related to Additional Lights Shadow Fade](https://github.com/Unity-Technologies/Graphics/pull/1616/commits/befe7f19a054c8a00e26e1b6760458eb910117ef).

**Comment by lukaschod** [comment]
> Thanks for unifying the interfaces.
> 
> The refactoring for AdditionalLights look good.
> Did you check that the addition of new uniforms in Shadows.hlsl does not make shaders exceed the low GL_MAX_FRAGMENT_UNIFORM_VECTORS limit we had in some of the SHADER_API_GLES30 devices (could be as low as 224 on Adreno3xx device I think)?
> 
> I'm also adding @Verasl, who instigated previous [fixes related to Additional Lights Shadow Fade](https://github.com/Unity-Technologies/Graphics/pull/1616/commits/befe7f19a054c8a00e26e1b6760458eb910117ef).

How is it some static check GL_MAX_FRAGMENT_UNIFORM_VECTORS ?

File: com.unity.render-pipelines.core/Editor/Lighting/Shadow/ShadowCascadeGUI.cs [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+                        borderWidth = 0;
+
+                    // Draw border snatch handle
+                    var borderPartitionHandleRect = new Rect(
+                        currentX - 6 - partitionHalfWidth,
+                        usableRect.y + usableRect.height - 1,
+                        12,
+                        18);
+                    var enabled = cascade.borderHandleState == HandleState.Enabled;
+                    var borderPartitionColor = enabled ? kCascadeColors[i] : kDisabledColor;;
```
Comment (martint-unity) [inline com.unity.render-pipelines.core/Editor/Lighting/Shadow/ShadowCascadeGUI.cs:228]: 2 :: at the end of this line
Nitpicky I know :)

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 7]
```diff
@@ -343,13 +346,38 @@ half AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS, half3 ligh
...
 }
 
-half GetShadowFade(float3 positionWS)
+half GetMainLightShadowFade(float3 positionWS)
+{
+    // This code computes distance between capsule and point.
+    // Finally distance is used with shadow fade scale and bias to compute shadow fade blend value.
+
+    float3 shadowFadeCenter = _CascadeShadowSplitSphereLast.xyz;
+    float3 worldSpaceCameraOppositeDir = -UNITY_MATRIX_V[2].xyz; // Move it to other property for faster access?
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:355]: Use GetWorldSpaceViewDir instead https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl#L94
 
 It's more efficient in case of perspective projection.

**Review by phi-lira — CHANGES_REQUESTED** [review]

**Comment by phi-lira** [comment]
Adding @martint-unity for awareness on the editor / cascade UI.

**Review by martint-unity — CHANGES_REQUESTED** [review]
I tested this PR on my machine. 
Our own tests are failing scenes 135-138. Seems like they default to 4 cascades so some sort of an upgrader is needed I guess otherwise I think we break users.
Also changing the cascade count to 1 throws an error. 
image

Please use our test scenes for this PR.

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 8]
```diff
@@ -343,13 +346,38 @@ half AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS, half3 ligh
...
+    float3 lineToPoint = positionWS - lineCenter;
+
+    // Find shortest vector between line and point
+    float distanceFromLineCenter = min(dot(lineDir, lineToPoint), 0);
+    float3 offsetFromLineCenter = lineDir * distanceFromLineCenter;
+    float3 shortestLineToPoint = lineToPoint - offsetFromLineCenter;
+
+    float distanceSq = dot(shortestLineToPoint, shortestLineToPoint);
+    half fade = saturate(distanceSq * _MainLightShadowParams.z + _MainLightShadowParams.w);
+    return fade;
```
Comment (nigeljw-unity) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:371]: This needs to be the min between the edge of the sphere, the shadow distance, and the far plane for the cascade.

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 9]
```diff
@@ -343,13 +346,38 @@ half AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS, half3 ligh
     return SampleShadowmap(TEXTURE2D_ARGS(_AdditionalLightsShadowmapTexture, sampler_AdditionalLightsShadowmapTexture), shadowCoord, shadowSamplingData, shadowParams, true);
 }
 
-half GetShadowFade(float3 positionWS)
+half GetMainLightShadowFade(float3 positionWS)
+{
+    // This code computes distance between capsule and point.
+    // Finally distance is used with shadow fade scale and bias to compute shadow fade blend value.
+
+    float3 shadowFadeCenter = _CascadeShadowSplitSphereLast.xyz;
```
Comment (nigeljw-unity) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:354]: We need to fade when the shadow gets clipped for all cascades. The shadow can be clipped by the sphere in any cascade level with the current implementation in trunk (this is really a bug with the original stable fit implementation but we intend to keep this functionality going forward for performance/quality reasons).

**Review by RSlysz — APPROVED** [review]
Seams good. New handles look nice !

Though you end up having same issue I had when writing the widget. Spaghetti function is hard to follow. If you find a way to split it into sub functions, it can increase readability. (DrawCascades method is 200+ lines of code)

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 10]
```diff
@@ -659,6 +660,14 @@ public Vector3 cascade4Split
             get { return m_Cascade4Split; }
         }
 
+        /// <summary>
+        /// Returns last cascade fade distance in procentage.
+        /// </summary>
+        public float cascadeBorder
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:666]: Not really I just followed it from cascade

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs [hunk 11]
```diff
@@ -202,6 +202,12 @@ public bool IsCameraProjectionMatrixFlipped()
         public int mainLightShadowmapHeight;
         public int mainLightShadowCascadesCount;
         public Vector3 mainLightShadowCascadesSplit;
+        /// <summary>
+        /// Main light last cascade shadow fade border.
+        /// Value represents the width of shadow fade that ranges from 0 to 1.
+        /// Where value 0 is used for no shadow fade.
+        /// </summary>
+        public float mainLightShadowCascadeBorder;
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:210]: Yes and yes. My though behind this was - it is safer to deprecate it later instead of introducing generic property which has empty values. Should I change it to vector4, not sure about the naming :/?

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 9 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:354]: Could you eleborate on this more, I am not sure I got the issue

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 8 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:371]: the last cascade should already be the minimum of shadow distance, far plane and of course sphere (As it is). Maybe I missunderstood here?

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 9 above)
Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:354]: The general problem you are fixing isn't specific to only the last cascade. Every cascade sphere clips the shadow in the corners.

**Comment by lukaschod** [comment]
> Seams good. New handles look nice !
> 
> Though you end up having same issue I had when writing the widget. Spaghetti function is hard to follow. If you find a way to split it into sub functions, it can increase readability. (DrawCascades method is 200+ lines of code)

Thanks, I actually wanted to increase readability. Going to see if I can create solid simple sub functions, wanted avoid doing as usually it makes things even worse if functions are not well though :D

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 8 above)
Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:371]: Unfortunately it is a false assumption. Stable fit was never implemented properly which gives this false impression. The implementation does try to do this, but it doesn't work, though the cascade frustum should actually be entirely within each sphere. 

image

Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:371]: The assumption is only true for square aspect ratios.

**Comment by Wilfrid-Unity** [comment]
> > Did you check that the addition of new uniforms in Shadows.hlsl does not make shaders exceed the low GL_MAX_FRAGMENT_UNIFORM_VECTORS limit we had in some of the SHADER_API_GLES30 devices (could be as low as 224 on Adreno3xx device I think)?
> How, is it some static check GL_MAX_FRAGMENT_UNIFORM_VECTORS ?

I think the easiest is to test on a real device, if there is an issue with a shader, "adb logcat" will contain something like "Unity   : Error: uniform variables in fragment shader do not fit in 224 vectors.". (I used a GalaxyS5(SC-04F) the last time I had to check that I did not hit that issue).

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 8 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:371]: Mhh, is it in all cases? As I did not noticed visually this effect. If it is rare case and will be fixed with ure internal change, maybe we can leave it as it is. Also HDRP should have same issue, as they also do shadow fade from cascade.

File: com.unity.render-pipelines.core/Editor/Lighting/Shadow/ShadowCascadeGUI.cs (hunk 6 above)
Comment (lukaschod) [reply com.unity.render-pipelines.core/Editor/Lighting/Shadow/ShadowCascadeGUI.cs]: Fixed it, we actually have quite a lot of them in code base :D. We should probably clean it up in some quality push

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 7 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:355]: Looking at code, it seems like it is going to be a bit heavier.

// Computes the world space view direction (pointing towards the viewer).
float3 GetWorldSpaceViewDir(float3 positionWS)
{
    if (IsPerspectiveProjection())
    {
        // Perspective
        return GetCurrentViewPosition() - positionWS;
    }
    else
    {
        // Orthographic
        return -GetViewForwardDir();
    }
}

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 8 above)
Comment (nigeljw-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:371]: I don't think it is a rare issue and it should be very common. These types of shadow problems are creating a lot of frustrations with users. With my change, this problem wouldn't happen for all cascades, so this fix you have made would not be necessary, but @phi-lira would like us to keep this functionality as a performance/quality optimization (and the default behavior still needs to be discussed). Yeah, HDRP has the same issue, but we should definitely fix it for URP. HDRP use the cascade sphere for a few different things.

**Review by Verasl — CHANGES_REQUESTED** [review]
Issues with popping when the shadows lie in the blend range, seems like the overall shadow casting distance is lowered when the blend is there which means things are popping out of being rendered in the shadow map.

https://user-images.githubusercontent.com/9811576/105481074-d3a40380-5ca6-11eb-93e0-983483c381e4.mp4

UX wise is quite nice, I think the end blend shouldnt change when changing in-between cascades, this is kind of annoying:
sMgTc9JAMZ

Other than that I like it very much and want it :D

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 8 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:354]: After offline discusion with Nigle we decided to revert shadow fade from last cascade. As this will be fixed by Nigle stable fit and current fix does not handle case described in here. HDRP is also suffering from that, so I think we should not follow their path here.
hdrp-issue

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:666]: I added setter

**Comment by lukaschod** [comment]
> Issues with popping when the shadows lie in the blend range, seems like the overall shadow casting distance is lowered when the blend is there which means things are popping out of being rendered in the shadow map.
> 
>  wnaOa5bcUd.mp4 
> UX wise is quite nice, switching between percentage and metric is not 'undo-able' which owuld be nice even though it's purely UI, also I think the end blend shouldnt change when changing in-between cascades, this is kind of annoying:
> sMgTc9JAMZ
> 
> Other than that I like it very much and want it :D

This should be fixed in latest commit.

>UI, also I think the end blend shouldnt change when changing in-between cascades, this is kind of annoying
This is from HDRP, maybe we can change it if it really a bad control

**Comment by lukaschod** [comment]
I just reverted shadow fading from last cascade, because it can not really be done for current stable fit solution without sacrficing performance. I will leave this to nigle stable fit fix.

**Review by Verasl — APPROVED** [review]
Shadow popping at end cascade fade gone.
LGTM

**Review by martint-unity — APPROVED** [review]

**Review by phi-lira — APPROVED** [review]

**Review by Wilfrid-Unity — APPROVED** [review]

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Issues I found seem to have been fixed, and there appear to be no performance issues on PC.
[Test doc](https://docs.google.com/document/d/1ZOqbY7BAIHJh6TwwKb8yaifn9Oc_zLcm04Q3p3Pz30w/edit#) here

**Review by Nzollner — APPROVED** [review]

**Review by ernestasKupciunas — APPROVED** [review]
Looks good on the mobile side.
I have found some bugs, but none of them were related to the changes (reproducible on Trunk and Master branches).

Testing doc: https://docs.google.com/document/d/12ZSHCzSR4j5ZW59EzOSyEZCSRSmPrsmYPjZvTWIs0Uo/edit# 
Performance sheet: https://docs.google.com/spreadsheets/d/1Xp_OZcZsY14dpY5P9pAviGCrhzKOT6lzI17bT9KL8yg/edit#gid=0

Devices under testing:
VLNQA00217, Razer Phone 2, 9.0.0, Snapdragon 845 SDM845, Adreno 630
VLNQA00268, Samsung Galaxy S10+, 10.0.0, Exynos 9 9820, Mali-G76
OnePlus Nord, 10, Snapdragon 765G SM7250-AB, Adreno 620
VLNQA00108, LGE LG Leon 4G LTE, 5.0.1, Snapdragon 410 MSM8916, Adreno 306
VLNQA00126, Samsung Galaxy S5 Mini, 6.0.1, Exynos 3 Quad 3470, Mali-400 MP
VLNQA00292, iPad Air3, SoC: A12, iOS: 13.2.3
iPad Air4, SoC: 14, iOS: 14.0
iPhone 12Pro, SoC: 14, iOS: 14.1
iPhone XS, SoC: A12, iOS: 14.3

**Comment by lukaschod** [comment]
I went and compared manually every failure that is not dependency against master. I was looking for similar failure not exectly on same base branch as some of the failures are instabilities. Seems like I did not introduced any new failures.

Build HDRP on Linux_Vulkan_mono_Linear_Standalone_build_Player on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-linux-vulkan.yml%2523Build_HDRP_Linux_Vulkan_Standalone_mono_Linear_trunk/5272259/job

Build ShaderGraph on Linux_Vulkan_mono_Linear_Standalone_cache_build_Player
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fshadergraph-linux-vulkan.yml%2523Build_ShaderGraph_Linux_Vulkan_Standalone_cache_mono_Linear_trunk/5273373/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fshadergraph-linux-vulkan.yml%2523Build_ShaderGraph_Linux_Vulkan_Standalone_cache_mono_Linear_trunk/5273420/job

Build Universal_Stereo on Win__mono_Linear_Standalone_cache_build_Player on version
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Funiversal_stereo-win.yml%2523Build_Universal_Stereo_Win_Standalone_cache_mono_Linear_trunk/5273360/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Funiversal_stereo-win.yml%2523Build_Universal_Stereo_Win_Standalone_cache_mono_Linear_trunk/5289254/job

Build VFX_URP on Win_DX11_mono_Linear_Standalone_cache_build_Player
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fvfx_urp-win-dx11.yml%2523Build_VFX_URP_Win_DX11_Standalone_cache_mono_Linear_trunk/5273367/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fvfx_urp-win-dx11.yml%2523Build_VFX_URP_Win_DX11_Standalone_cache_mono_Linear_trunk/5256564/job

Formatting
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252F_formatting.yml%2523formatting/5273406/job
vfx missing

HDRP on Linux_Vulkan_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp-linux-vulkan.yml%2523HDRP_Linux_Vulkan_playmode_cache_mono_Linear_trunk/5273394/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-linux-vulkan.yml%2523HDRP_Linux_Vulkan_playmode_cache_mono_Linear_trunk/5289316/job

HDRP on OSX_Metal_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp-osx-metal.yml%2523HDRP_OSX_Metal_playmode_cache_mono_Linear_trunk/5273391/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-osx-metal.yml%2523HDRP_OSX_Metal_playmode_cache_mono_Linear_trunk/5267775/job

HDRP on Win_DX11_Standalone_cache_mono_Linear on version trunk
https://yamato-artifactviewer.prd.cds.internal.unity3d.com/cb0e627a-b05d-428c-a2c9-b6331edba48b%2Flogs%2FTestProjects%2FHDRP_RuntimeTests%2Ftest-results/TestReportV1.html
System.AggregateException: One or more errors occurred. (The system cannot find the file specified.)
 ---> System.ComponentModel.Win32Exception (2): The system cannot find the file specified.
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)

HDRP on Win_DX11_playmode_XR_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp-win-dx11.yml%2523HDRP_Win_DX11_playmode_XR_cache_mono_Linear_trunk/5273387/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-win-dx11.yml%2523HDRP_Win_DX11_playmode_XR_cache_mono_Linear_trunk/5272254/job

HDRP on Win_DX11_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp-win-dx11.yml%2523HDRP_Win_DX11_playmode_cache_mono_Linear_trunk/5273386/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-win-dx11.yml%2523HDRP_Win_DX11_playmode_cache_mono_Linear_trunk/5272253/job

HDRP on Win_DX12_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp-win-dx12.yml%2523HDRP_Win_DX12_playmode_cache_mono_Linear_trunk/5273389/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-win-dx12.yml%2523HDRP_Win_DX12_playmode_cache_mono_Linear_trunk/5272256/job

HDRP on Win_Vulkan_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp-win-vulkan.yml%2523HDRP_Win_Vulkan_playmode_cache_mono_Linear_trunk/5273390/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp-win-vulkan.yml%2523HDRP_Win_Vulkan_playmode_cache_mono_Linear_trunk/5267774/job

HDRP_Hybrid on Win_DX11_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fhdrp_hybrid-win-dx11.yml%2523HDRP_Hybrid_Win_DX11_playmode_cache_mono_Linear_trunk/5273397/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fhdrp_hybrid-win-dx11.yml%2523HDRP_Hybrid_Win_DX11_playmode_cache_mono_Linear_trunk/5272264/job

Universal on Android_OpenGLES3_Standalone_cache_il2cpp_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Funiversal-android-opengles3.yml%2523Universal_Android_OpenGLES3_Standalone_cache_il2cpp_Linear_trunk/5273353/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Funiversal-android-opengles3.yml%2523Universal_Android_OpenGLES3_Standalone_cache_il2cpp_Linear_trunk/5271296/job

Universal on Android_Vulkan_Standalone_cache_il2cpp_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Funiversal-android-vulkan.yml%2523Universal_Android_Vulkan_Standalone_cache_il2cpp_Linear_trunk/5273355/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Funiversal-android-vulkan.yml%2523Universal_Android_Vulkan_Standalone_cache_il2cpp_Linear_trunk/5267711/job

Universal_Hybrid on OSX_Metal_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Funiversal_hybrid-osx-metal.yml%2523Universal_Hybrid_OSX_Metal_playmode_cache_mono_Linear_trunk/5273365/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Funiversal_hybrid-osx-metal.yml%2523Universal_Hybrid_OSX_Metal_playmode_cache_mono_Linear_trunk/5267721/job

Universal_Hybrid on Win_DX11_playmode_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Funiversal_hybrid-win-dx11.yml%2523Universal_Hybrid_Win_DX11_playmode_cache_mono_Linear_trunk/5273364/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Funiversal_hybrid-win-dx11.yml%2523Universal_Hybrid_Win_DX11_playmode_cache_mono_Linear_trunk/5271298/job

VFX_HDRP on Win_DX11_Standalone_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fvfx_hdrp-win-dx11.yml%2523VFX_HDRP_Win_DX11_Standalone_cache_mono_Linear_trunk/5273400/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fvfx_hdrp-win-dx11.yml%2523VFX_HDRP_Win_DX11_Standalone_cache_mono_Linear_trunk/5272267/job

VFX_HDRP on Win_DX12_Standalone_cache_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fshadow-fade-improvements/.yamato%252Fvfx_hdrp-win-dx12.yml%2523VFX_HDRP_Win_DX12_Standalone_cache_mono_Linear_trunk/5273403/job
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fvfx_hdrp-win-dx12.yml%2523VFX_HDRP_Win_DX12_Standalone_cache_mono_Linear_trunk/5272270/job

**Comment by PaulDemeulenaere** [comment]
I guess you already noticed but this merge request is leading to a missing `.meta`  due to the addition of the folder `com.unity.render-pipelines.core\Editor\Lighting\Shadow`.
It isn't a big deal but I think it will make fail the package verification.

**Comment by theo-pnv** [comment]
^ PR: https://github.com/Unity-Technologies/Graphics/pull/3480
